Implement keyed messages

alpreu opened the pull request:
This PR should implement keyed messaging on the PulsarTemplate (#26). 
A detailed description follows.

onobc commented on spring-pulsar/src/main/java/org/springframework/pulsar/core/PulsarOperations.java:
```diff
@@ -71,7 +71,20 @@ default MessageId send(T message, MessageRouter messageRouter) throws PulsarClie
+	 * @return the id of the sent message
+	 * @throws PulsarClientException if an error occurs
+	 */
+	MessageId send(String topic, T message, MessageRouter messageRouter, TypedMessageBuilderCustomizer<T> typedMessageBuilderCustomizer) throws PulsarClientException;
```
Wdyt about adding a few more combos to accommodate use of customizer w/o router (and vice-versa)?

**Current combos:**
```
- message
- message / topic
- message / router 
- message / topic / router
```
**New combos something like this:** 
``` 
- message / customizer
- message / topic / customizer
- message / customizer / router
- message / topic / customizer / router
```
The tradeoff of course is signature explosion vs. user convenience. For me, the concerns of signature explosion is lightened by the `default impls delegate to the single abstract w/ all of the params` pattern we are using.

The other tiny change I think we can make it put the customizer in front of the router in the arg list as I bet router is used less than customizer and the customizer would come directly after the payload arg (cohesive).

alpreu replied:
Thanks for the suggestions, and yes, having the customizer together with the payload makes sense 👍

onobc commented on spring-pulsar/src/main/java/org/springframework/pulsar/core/TypedMessageBuilderCustomizer.java:
```diff
@@ -0,0 +1,30 @@
+import org.apache.pulsar.client.api.TypedMessageBuilder;
+
+/**
+ * The function to customize a {@link TypedMessageBuilder}.
```
I used an actual function in the backing issue, but that was mostly for illustration. I think we can use something similar to what Boot uses such as [here](https://github.com/spring-projects/spring-boot/blob/main/spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/kafka/DefaultKafkaProducerFactoryCustomizer.java). 

The 2 important things in there are 
* `@FunctionalInterface` marker on the interface
* Signature change slightly to `void customize(TheThingWeAreCustomizingHere<?, ?> theThing)`

onobc reviewed (changes requested):
That was fast @alpreu ! Thanks, looks great. I have a few minor comments/suggestions to follow up on. Additionally,

1. Please add yourself as `@author` on all modified files. 
2. Do you mind adding some tests cases to the existing PulsarTemplateTests?

Thanks again for the contribution!

onobc commented on spring-pulsar/src/main/java/org/springframework/pulsar/core/TypedMessageBuilderCustomizer.java:
```diff
@@ -0,0 +1,34 @@
+ * @author Alexander Preuß
+ */
+@FunctionalInterface
+public interface TypedMessageBuilderCustomizer {
```
Wdyt about adding the type `<T>` here and parameterize the parameter below w/ it?

onobc reviewed (changes requested):
Looks great @alpreu . Thanks for the updates and adding the tests. The only comment I have is around type info to the customizer. Once we resolve that we can remove the draft status and I will get it merged in.

alpreu commented:
Thank you for the feedback, I added the type to the message builder and updated the docs to reflect the changes. If it looks good to you I'll squash and rebase so you can merge :)

onobc commented:
Looks good @alpreu . I went ahead and squash/rebase/merge. Thanks again for contribution!
